### ISSUE TITLE
Least Squares optimization in Picos

### ISSUE THREAD

**Issue by BEpresent** [gsagnol/picos#16]
Dear all,

 I tried to solve a simple least squares problem without constraints. ||Ax -b||**2

Given are mxn matrix A and target-vector b with length m, and I just want to calculate the optimal x. This can be easily done in Numpy and Scipy (in fact A and b are Numpy arrays), however, I want to understand the Picos syntax for this simple case before I add quadratic constraints in the next step. Gurobi is installed as an additional solver.

import picos as pic
import cvxopt as cvx
P = pic.Problem()
w = P.add_variable('w',(Nc*Nspokes), 'complex')

A = cvx.matrix(Afull)
b = cvx.matrix(target)
A = pic.new_param('A',A)
b = pic.new_param('b',target)
P.set_objective('min', abs(A_w -b)_*2)

P.solve(verbose=1)

I get

 Solved in 0 iterations and 0.00 seconds
Optimal objective  0.000000000e+00
**\* Primal Solution not found

Thank you

**Comment by gsagnol** [comment]
Hi,

do you really need the variable w to be complex? Are A and b complex indeed ?
Otherwise, you should write:

w = P.add_variable('w',(Nc*Nspokes))

Then, the set_objective line should be

P.set_objective('min', abs(A_w -b)_*2)

**Comment by gsagnol** [comment]
Sorry, without code environment some multiply signs (stars) disappeared:

`P.set_objective('min', abs(A*w -b)**2)`

**Comment by BEpresent** [comment]
Hi, A and the target-vector b are indeed complex,  Nc*Nspokes ist just a number, i.e. 8. I require the solution to be a complex vector.

**Comment by gsagnol** [comment]
Can you please tell me the output of
`w.value = cvx.matrix([1,2,3,4,5,6,7-1j,8+2j])`
`print abs(A*w -b)`
`print (A*w).H`

**Comment by BEpresent** [comment]
abs(A_w -b)
 norm of a (2321 x 1)- expression: ||A_w -b|| 

print(abs(A_w -b))
 norm of a (2321 x 1)- expression: ||A_w -b|| 

print ((A_w).H)
 (1 x 2321)-affine expression: ( A_w ).H 

print(abs(A_w -b)__2)
quadratic expression: ||A_w -b||**2 

 w
  (40 x 1)-affine expression: w 

when I try w.value I get 

Exception: w_IM is not valued

**Comment by gsagnol** [comment]
OK, This might be a bug I had corrected but not released yet. Try to download the latest development version directly from github. If this does not work, send me a code with a minimal example (take a smaller matrix A.)

**Comment by BEpresent** [comment]
Picos 1.1.1 is the most recent version, isn't it? 

Could it be that Gurobi  is not recognized  correctly? If I try in the terminal python picos/test_picos.py it gives me a lot of "failed" messages, however, cvxopt works fine. 

In IPython Gurobi seems to be recognized.

Here is a small example with a 10x10 matrix and a size 10 target vector, both complex:

import numpy as np

A = np.random.rand(10, 10)  + 1j*np.random.rand(10, 10)

b = np.random.rand(10)  + 1j*np.random.rand(10)

import picos as pic
import cvxopt as cvx

P = pic.Problem()
w = P.add_variable('w',(10),'complex')
A = cvx.matrix(A)
b = cvx.matrix(b)

A = pic.new_param('A',A)
b = pic.new_param('b',b)

P.set_objective('min', abs(A_w -b)_*2)

Sometimes I get errors with P.solve(verbose=1)

then I try P.solve(verbose=0) or P.solve()

**Comment by gsagnol** [comment]
Yes, picos 1.1.1. is the most recent release, but you can download the 1.1.2.dev on github.

I think that quadratics with complex variables are supported in constraints only, but not in the objective function.

So adding an auxiliary variable should solve your problem:
`P.add_variable('t',1)`
`P.add_constraint(abs(A*w-b)**2<t)`
`P.minimize(t)` (or `P.set_objective('min',t);P.solve()` )

Tell me if this works.

**Comment by gsagnol** [comment]
For gurobi, check in the  console if `import gurobipy` throws some error ?

**Comment by BEpresent** [comment]
`import gurobipy` does work without problems. 

Your solution gives me a result, unfortunately it does not solve the least-squares problem correctly. (I get a lot of zero entries in my solution vector).

Where I want to get at is to have a solution to the least-squares problem stated above with quadratic inequality constraints. I wonder if my solvers are even suited for that or if I should install other solvers? Efficient primal-dual interior point methods would be preferred. Do you have suggestions for that?

Thank you very much

**Comment by gsagnol** [comment]
Hi,

that' funny. So here is an example with a seeded random generator to 
make reproducible results.

``` python
import numpy as np
n_vars = 5
n_eq = 8
numpy.random.seed(42) #seed the random generator
A = np.random.rand(n_eq, n_vars) + 1j*np.random.rand(n_eq, n_vars)
b = np.random.rand(n_eq) + 1j*np.random.rand(n_eq)
import picos as pic
import cvxopt as cvx
P = pic.Problem()
w = P.add_variable('w',n_vars,'complex')
t = P.add_variable('t',1)
A = cvx.matrix(A)
b = cvx.matrix(b)
A = pic.new_param('A',A)
b = pic.new_param('b',b)
P.add_constraint(abs(A*w-b)**2 < t)
P.set_objective('min', t)
P.solve(solver='gurobi')
```

I get an optimal solution without zeroes. By the way, what PICOS does 
internally is just transforming the 'complex-valued problem'
into the following 'real-valued' problem:

``` python
numpy.random.seed(42) #seed the random generator
Ar = pic.new_param('Ar',cvx.matrix(np.random.rand(n_eq, n_vars)))
Ac = pic.new_param('Ac',cvx.matrix(np.random.rand(n_eq, n_vars)))
br = pic.new_param('br',cvx.matrix(np.random.rand(n_eq)))
bc = pic.new_param('bc',cvx.matrix(np.random.rand(n_eq)))

P2 = pic.Problem()
wr = P2.add_variable('wr',n_vars)
wc = P2.add_variable('wc',n_vars)
tr = P2.add_variable('tr',1)
tc = P2.add_variable('tc',1)

P2.add_constraint(abs(Ar*wr-Ac*wc-br)**2<tr)
P2.add_constraint(abs(Ar*wc+Ac*wr-bc)**2<tc)
P2.minimize(tr+tc)
```

I get the same solution for both P and P2, namely an objective value of 
`0.76617872`
and `w`is equal to
[ 3.72e-02+j4.53e-01]
[ 9.36e-02-j4.99e-01]
[ 5.77e-02+j1.13e-01]
[ 6.91e-01+j1.90e-02]
[ 2.31e-01-j1.31e-01]

Am 14.02.2016 um 18:02 schrieb BEpresent:

**Comment by BEpresent** [comment]
Interesting,

Indeed your example works, Thank you!

Now if I add quadratic constraints

```
A = np.random.rand(2300,8) + 1j*np.random.rand(2300,8)
b = np.random.rand(2300) + 1j*np.random.rand(2300)

P = pic.Problem()
w = P.add_variable('w',8,'complex')
t = P.add_variable('t',1)
A = cvx.matrix(A)
b = cvx.matrix(b)
A = pic.new_param('A',A)
b = pic.new_param('b',b)

P.add_constraint(abs(A*w-b)**2 < t)
```

Take this positive semidefinite (psd) matrix Q

```
Q = np.array([[  1.74636825e-02 +3.28621814e-13j,
         -4.75287484e-03 -5.23655443e-03j,
          3.29191447e-03 +4.03854996e-03j,
         -8.52801546e-04 +4.50710009e-04j,
          3.51163733e-04 -5.18687710e-04j,
          4.88034362e-04 +1.64259502e-04j,
          5.93758632e-05 -4.73097549e-04j,
         -4.48899547e-04 +3.38563579e-04j],
       [ -4.75287484e-03 +5.23655443e-03j,
          4.39293738e-02 -4.66813957e-12j,
         -2.11890042e-02 -1.56821217e-03j,
          1.23240426e-03 -4.13039792e-03j,
          8.23433453e-04 +2.46473658e-03j,
         -1.38162530e-03 +9.09156806e-04j,
          1.44044170e-03 +1.46799441e-03j,
          6.24255335e-04 -1.69300847e-03j],
       [  3.29191494e-03 -4.03854996e-03j,
         -2.11890042e-02 +1.56821229e-03j,
          3.93388588e-02 -2.66580681e-12j,
         -7.92735373e-04 +4.90170671e-03j,
          1.55409827e-04 -1.17120787e-03j,
          1.74759096e-03 +5.62859816e-04j,
         -1.04194868e-03 +2.55709310e-04j,
          2.14358151e-04 +2.46560713e-03j],
       [ -8.52801546e-04 -4.50710009e-04j,
          1.23240426e-03 +4.13039792e-03j,
         -7.92735489e-04 -4.90170671e-03j,
          1.67827193e-02 +1.18722415e-13j,
         -1.90346633e-04 +1.70350599e-04j,
          1.05859654e-04 -3.45926441e-04j,
          1.86171834e-04 +2.98599334e-04j,
          5.60273009e-04 -7.26833387e-05j],
       [  3.51163762e-04 +5.18687826e-04j,
          8.23433453e-04 -2.46473658e-03j,
          1.55409827e-04 +1.17120775e-03j,
         -1.90346633e-04 -1.70350613e-04j,
          1.56905486e-02 +2.24784316e-14j,
          2.33412371e-04 +2.22427872e-04j,
          2.84483278e-04 +1.79514227e-05j,
          3.38429418e-05 -2.15439413e-05j],
       [  4.88034362e-04 -1.64259502e-04j,
         -1.38162530e-03 -9.09156748e-04j,
          1.74759096e-03 -5.62859816e-04j,
          1.05859668e-04 +3.45926441e-04j,
          2.33412371e-04 -2.22427872e-04j,
          1.56376120e-02 +1.15504516e-13j,
          1.88000951e-04 -1.37567931e-05j,
          1.66987666e-04 +1.14629984e-04j],
       [  5.93758632e-05 +4.73097520e-04j,
          1.44044159e-03 -1.46799441e-03j,
         -1.04194868e-03 -2.55709310e-04j,
          1.86171834e-04 -2.98599305e-04j,
          2.84483278e-04 -1.79514263e-05j,
          1.88000937e-04 +1.37567949e-05j,
          1.56079930e-02 -2.22785373e-13j,
          1.54855952e-04 -1.95474815e-04j],
       [ -4.48899547e-04 -3.38563579e-04j,
          6.24255335e-04 +1.69300847e-03j,
          2.14358151e-04 -2.46560713e-03j,
          5.60272951e-04 +7.26833168e-05j,
          3.38429454e-05 +2.15439413e-05j,
          1.66987666e-04 -1.14629998e-04j,
          1.54855967e-04 +1.95474815e-04j,
          1.56488115e-02 +1.25290037e-13j]])



Q = cvx.matrix(Q)
Q = pic.new_param('Q',Q)
P.add_constraint(w.H*Q*w -3.0 < 0)
P.solve(solver='gurobi')
```

Therefore I have the problem

```
optimization problem  (Mixed (SOCP+quad)):
17 variables, 0 affine constraints, 2302 vars in 1 SO cones, 256 nnz  in 1 quad constraints

w_IM    : (8, 1), continuous
t       : (1, 1), continuous
w_RE    : (8, 1), continuous

        find vars
such that
  ||A*w -b||^2 < t
  w.H*Q*w -0 < 0
```

The problem is that for any complex w

`w.H*Q*w = (0.082661925254814861+5.1953233382029396e-11j)
`
i.e. there is a small numerical imprecision  +5.1953233382029396e-11j which is essential = 0 and therefore Q is indeed psd.

Therefore I get the error 

```
P.solve(solver='gurobi')

AttributeError: 'QuadExp' object has no attribute 'is_real'
```

In Python I can easily make this real like with abs(w.H_Q_w), however in Picos this is not allowed. Is there a way to make w.H_Q_w real (without imprecision?)

**Comment by gsagnol** [comment]
You're right, complex quadratics are not completely supported yet. Meanwhile, there is an easy workaround:

First, you have to force Q to be Hermitian indeed:
`Q = 0.5*(Q+np.conj(Q).T)`

Then, take a Cholesky decomposition of Q, so that Q = LL.H

```
L = np.linalg.cholesky(Q)
L = cvx.matrix(L)
L = pic.new_param('L',L)
```

Finally, your constraint is equivalent to the following SOCP constraint:
`P.add_constraint( abs(L.H*w)**2 < 3)`

**Comment by BEpresent** [comment]
This works,very nice, thank you.

I have an additional constraint, `P.add_constraint(abs(w*w)**2 < 0.5),`

however, I get 

```
File "/opt/anaconda/lib/python3.5/site-packages/picos/expression.py", line 938, in __mul__
    raise Exception('not implemented')

Exception: not implemented

```

How can I add a constraint like this in addition to the others from above?

**Comment by gsagnol** [comment]
What does `w*w` mean ? 
If you mean `w.H*w`, then it is the same as `abs(w)**2 < 0.5`
But  `abs(w.H*w)**2` would be a polynomial of degree 4 in `w`, I doubt this is what you want to do ?

**Comment by BEpresent** [comment]
You are correct, I reformulated my w internally.

Let's say I have a solution, then why is `abs(w)**2` still complex?

e.g.

```
print(abs(w)**2) 
(0.20553007242540822+0.025327221487294632j)
```

Let's say I also want to add constraints on the individual components/the sum thereof, like 

`P.add_constraint(abs(w[0])**2+abs(w[1])**2+abs(w[2])**2+abs(w[3])**2+abs(w[4])**2 < 0.1)
`
It accepts the constraint, however I get 

`AttributeError: 'QuadExp' object has no attribute 'is_real'
`
It seems, that internally the abs() is treated as a complex number?

```
 print(w[i].H*w[i])
(0.0009146383397344392+2.4552024999794734e-20j)

```

Thank you

**Comment by gsagnol** [comment]
As I said,
quadratics with complex variables are not handled yet in python, I will 
try to correct this in the next release.
However you could do:

|```
P.add_constraint(abs(w[0])**2 < t0
|||P.add_constraint(abs(w[1])**2 < t1
|||P.add_constraint(abs(w[2])**2 < t2
||P.add_constraint(abs(w[3])**2 < t3
|||P.add_constraint(abs(w[4])**2 < t4|
||||P.add_constraint(||t0+t1+t2+t3+t4 < 0.1)

```
||
|
||`abs(w)**2`| |returns a wrong value because of this bug with complex 
quadratics,
but internally when a constraint of the form `abs(...)**2 < ...`
is passed to picos, picos does the correct job with it. To see the true 
value of ||||||`abs(w)**2`|| |,
try to do
`print w.H * w`

Best,
Guillaume
||
|Am 19.02.2016 um 15:53 schrieb BEpresent:
>
> You are correct, I reformulated my w internally.
>
> Let's say I have a solution, then why is |abs(w)**2| still complex?
>
> e.g.
>
> |print(abs(w)**2) (0.20553007242540822+0.025327221487294632j) |
>
> Let's say I also want to add constraints on the individual 
> components/the sum thereof, like
>
> |P.add_constraint(abs(w[0])**2+abs(w[1])**2+abs(w[2])**2+abs(w[3])**2+abs(w[4])**2 
> < 0.1)
> |
> It accepts the constraint, however I get
>
> |AttributeError: 'QuadExp' object has no attribute 'is_real'
> |
> It seems, that internally the
>
> pic.abs still treats the abs() as a complex number?
>
> Thank you
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/gsagnol/picos/issues/16#issuecomment-186244202>.
>

```

**Comment by BEpresent** [comment]
It seems to work, Thank you very much for the help. 

Looking forward to the next update with quadratic complex constraints.

A question remains, however:

It says that the primal objective, is zero, i.e. in the end I get

```
                  Objective                Residual
Iter       Primal          Dual         Primal    Dual     Compl     Time

   7   0.00000000e+00 -3.09390705e-09  1.00e-10 9.14e-15  5.91e-13     0s
```

How is that possible? In Picos I never actually set a specific objective but only constraints. I assume it translates the constraints into an objective? Even then  the objective can be hardly zero?

The Primal residual would be something like Ax-b,  the actual value is far from being  1.00e-10, even if I take the norm.

**Comment by gsagnol** [comment]
Well, if you have only constraints, then it is merely a feasibility 
problem,
and picos consructs the problem of minimizing "0" subject to the 
constraints c1,...,cn.

I don't know anymore what exactly your problem look like. If you 
minimize `t` such that
`abs(A*x-b)**2 < t`, then the objective value should be the value of the 
residual `t`  indeed,
again, give me a minimal example so I can have a look at it.

Cheers,
Guillaume.

**Comment by BEpresent** [comment]
Here is some example code; its size is comparable to the problem I want to solve:

```
A = np.random.rand(4000,80)
b = np.random.rand(4000)
P = pic.Problem()
w = P.add_variable('w',80)
t = P.add_variable('t',1)
A = cvx.matrix(A)
b = cvx.matrix(b)
A = pic.new_param('A',A)
b = pic.new_param('b',b)

P.add_constraint(abs(A*w-b)**2 < t)
P.solve(solver='gurobi')
```

In addition I have more constraints, however, here I am interested in the output, especially the residuals.  Gurobi seems to use a barrier method to solve the problem. The output is something like 

```
                  Objective                Residual
Iter       Primal          Dual         Primal    Dual     Compl     Time
   0   0.00000000e+00  0.00000000e+00  1.25e+02 1.00e-01  9.72e-03     0s
   1   0.00000000e+00  1.33465280e+00  3.39e+00 1.16e-02  3.79e-04     0s
   2   0.00000000e+00  1.26428571e+00  2.16e+00 4.79e-03  7.22e-05     0s
   3   0.00000000e+00  1.58096233e+00  1.96e+00 3.36e-03  3.29e-05     1s
   4   0.00000000e+00  5.37709815e-01  8.47e-01 1.71e-03  5.28e-05     1s
   5   0.00000000e+00  9.71235537e-01  5.12e-01 1.54e-03  2.16e-04     1s
   6   0.00000000e+00 -3.95030163e-03  6.17e-07 1.56e-06  1.45e-06     1s
   7   0.00000000e+00 -3.96364667e-06  9.04e-09 1.57e-09  1.45e-09     1s
   8   0.00000000e+00 -3.97808770e-09  2.61e-10 1.57e-12  1.46e-12     1s

```

Calculating the norm of Ax-b gets me 22.89362 in one random example, wouldn't this correspond to the  Primal Residual? Something converges, but the  internal functions that converge are not obvious to me.

**Comment by gsagnol** [comment]
OK, so you don't optimize anything with your code !! (so the returned 
residual could be arbitrarily large, Gurobi just tries to find values of 
x and t satisfying `abs(A*x-b)**2<=t`)

you must add
`P.set_objective('min',t)`
before the call to `P.solve()|`, or
alternatively you can call the solver

directly with
  `P.minimize(t)`.

Best,
Guillaume.

|||

**Comment by BEpresent** [comment]
Indeed, looks much better :) Now the primal objective is consistent with norm(Ax-b)**2.

```
                  Objective                Residual
Iter       Primal          Dual         Primal    Dual     Compl     Time
   0   1.35315600e+02 -1.00200000e+00  1.36e+02 2.15e-01  5.66e-02     0s
   1   2.53216443e+01  5.39561901e+00  1.81e+01 2.15e-07  6.55e-03     0s
   2   2.73232198e+01  4.67612206e+01  1.15e+01 5.94e-10  6.16e-04     0s
   3   3.33045775e+01  5.12642352e+01  9.36e+00 5.53e-10  8.12e-04     1s
   4   3.43358746e+01  6.50461355e+01  8.96e+00 6.34e-09  8.63e-05     1s
   5   3.45097715e+01  6.63618139e+01  8.91e+00 6.13e-09  8.59e-05     1s
   6   3.45318634e+01  6.48868282e+01  8.90e+00 6.13e-09  8.74e-05     1s
   7   3.45333182e+01  6.69076118e+01  8.90e+00 6.12e-09  8.93e-05     1s
   8   3.45821157e+01  6.45725525e+01  8.89e+00 6.12e-09  9.27e-05     1s
   9   3.45871311e+01  6.76173445e+01  8.89e+00 6.11e-09  9.59e-05     1s
  10   3.46854020e+01  6.41935336e+01  8.87e+00 6.10e-09  1.02e-04     1s
  11   3.47005608e+01  6.84047150e+01  8.86e+00 6.09e-09  1.05e-04     1s
  12   3.48592707e+01  6.38452943e+01  8.83e+00 6.08e-09  1.14e-04     2s
  13   3.48972841e+01  6.91177427e+01  8.81e+00 6.04e-09  1.15e-04     2s
  14   3.50976972e+01  6.36595568e+01  8.77e+00 6.03e-09  1.24e-04     2s
  15   3.51734078e+01  6.96574519e+01  8.74e+00 5.98e-09  1.20e-04     2s
  16   3.53752572e+01  6.36967469e+01  8.69e+00 5.95e-09  1.29e-04     2s
  17   3.54975336e+01  7.00554839e+01  8.66e+00 5.89e-09  1.22e-04     2s
  18   3.56704531e+01  6.38719963e+01  8.61e+00 5.86e-09  1.30e-04     2s
  19   3.58450982e+01  7.04356992e+01  8.57e+00 5.83e-09  1.23e-04     2s
  20   3.59828517e+01  6.39811885e+01  8.53e+00 5.80e-09  1.34e-04     3s
  21   3.62346662e+01  7.09638601e+01  8.48e+00 5.76e-09  1.30e-04     3s
  22   3.63460197e+01  6.36850690e+01  8.43e+00 5.71e-09  1.51e-04     3s
  23   3.67819489e+01  7.19072407e+01  8.37e+00 5.68e-09  1.57e-04     3s
  24   3.68768283e+01  6.20592668e+01  8.31e+00 5.75e-09  2.22e-04     3s
  25   3.80730581e+01  7.40300752e+01  8.16e+00 5.67e-09  2.85e-04     3s
  26   3.84656027e+01  5.11561736e+01  7.97e+00 6.44e-09  1.02e-03     3s
  27   4.01024627e+01  6.68258422e+01  7.63e+00 5.91e-09  7.61e-04     3s
  28   7.26340391e+01  8.94138547e+01  4.06e+00 4.93e-09  2.68e-03     4s
  29   7.59249767e+01  1.40285303e+02  3.59e+00 4.15e-08  4.19e-04     4s
  30   8.03852426e+01  1.06048027e+02  3.29e+00 8.58e-09  2.45e-03     4s
  31   8.08831055e+01  1.51258180e+02  3.27e+00 8.60e-09  2.40e-03     4s
  32   2.96400860e+02  1.70071277e+02  1.63e+00 1.13e-08  4.60e-02     4s
  33   2.31214420e+02  2.71930784e+02  4.72e-01 1.91e-08  2.46e-03     4s
  34   2.83500527e+02  2.66258771e+02  2.36e-01 1.07e-08  1.10e-02     4s
  35   3.27592804e+02  2.67725262e+02  1.18e-01 5.44e-09  1.95e-02     4s
  36   3.36969233e+02  3.28223991e+02  4.11e-03 4.79e-10  2.46e-03     4s
  37   3.32735035e+02  3.32520499e+02  3.64e-06 1.32e-10  5.36e-05     4s
  38   3.32603847e+02  3.32603276e+02  1.03e-07 7.78e-11  1.43e-07     5s
  39   3.32603403e+02  3.32603402e+02  1.08e-09 1.59e-09  2.00e-10     5s

Barrier solved model in 39 iterations and 4.79 seconds
```

What then is the Primal Residual which I thought would be Ax-b ? It starts large and converges to zero.  Ax-b is of course a vector quantity, and even in the optimal case it has non-zero elements, i.e. the target b can not be achieved perfectly.
